rfcs/20230622-quantized-reduction.md

```diff
@@ -0,0 +1,372 @@
+        %add = "stablehlo.add"(%lhs, %rhs)
+            : (tensor<!quant.uniform<i32:f32, accum_scale:accum_zp>>,
+               tensor<!quant.uniform<i32:f32, accum_scale:accum_zp>>)
+            -> tensor<!quant.uniform<i32:f32, accum_scale:accum_zp>>
+        "stablehlo.return"(%add)
+            : (tensor<!quant.uniform<i32:f32, accum_scale:accum_zp>>) -> ()
+    }) {
+        dimensions = dense<1> : tensor<i64>
+    } : (tensor<5 x 1 x !quant.uniform<ui8:f32, input_scale:input_zp>>,
+         tensor<!quant.uniform<ui8:f32, input_scale:input_zp>>)
+    -> tensor<5 x !quant.uniform<ui8:f32, output_scale:output_zp>>
+```
+
+### Semantics
+
+If (1) the input operand type is different from the reduction block
+argument type or (2) the op result type is different from the reduction block
+return type, there will be implicit type conversion defined by either
+`stablehlo.convert`, `stablehlo.uniform_quantize`, or
+`stablehlo.uniform_dequantize`. For example,
+
+ | Implicit type conversion op       | element type of operand or result type | element type of block argument or block return type |
+ |-----------------------------------|----------------------------------------|-----------------------------------------------------|
+ | (A) `stablehlo.uniorm_quantize`   | quantized tensor                       | quantized tensor                                    |
+ | (B) `stablehlo.uniorm_quantize`   | floating point                         | quantized tensor                                    |
+ | (C) `stablehlo.uniorm_dequantize` | quantized tensor                       | floating point                                      |
+ | (D) `stablehlo.convert`           | floating-point                         | integer                                             |
+ | (E) `stablehlo.convert`           | integer                                | floating-point                                      |
+ | (F) `stablehlo.convert`           | floating-point                         | floating-point                                      |
+ | (G) `stablehlo.convert`           | integer                                | integer                                             |
+ | (H) `stablehlo.convert`           | complex                                | complex                                             |
+
+At this point there is no use for cases other than (A), (F), and (G).  My
+proposal here would be to address (A), (F), and (G) only.  Note that the (F)
+partially addresses
+[Decide on mixed precision](https://github.com/openxla/stablehlo/issues/369)
+for reduce op in that it allows the the input or init value to differ from the
+corresponding block arguments w.r.t the precision of floating-point types.
+However, the mixed precision implementation in HLO seems more detailed in the
+sense that even allows `inputs` and `init_values` to differ in floating-point
+precision. My proposal would be to treat the above ticket separately.
```

Comment:
Can you explain why this should be its own ticket? Is there a risk that we have to fundamentally revisit the encoding? It seems there is a natural extension to this case by using different type conversions. Is the question then whether this should be supported?
It would be useful to at least discuss whether the mixed precision usecase can be covered.

Reply:
The current proposal does allows mixed precision for reduce op, but it solved the issue partially because there are other ops (than reduce) which need support for mixed precision [list](https://github.com/tensorflow/tensorflow/blob/1d69ba72834b963b72075a82c10959f6bb74e473/tensorflow/compiler/xla/service/hlo_verifier.cc#L1681-L1714). Having said that  my proposal is to address all those ops as part of resolving #369.

>> It would be useful to at least discuss whether the mixed precision usecase can be covered.
I agree with your point. I will update the doc with how the current proposal partially solves #369 for reduce op and the plan ahead to address #369.